### ISSUE TITLE
Fix for CVE-2022-21658 is unsafe on some file systems

### ISSUE THREAD

**Issue by jclulow** [rust-lang/rust#94335]
The fix for CVE-2022-21658 (#93112, 54e22eb7dbb615bd44355028d3fd867aa93c0972) appears to have introduced a regression that can cause file system corruption on some UNIX systems.

The new code will attempt to use [unlinkat(2)](https://illumos.org/man/2/unlinkat) in any case where it does not know what type of file a directory entry represents.  Some systems do not provide that information along with the entry names returned while reading a directory, and even on systems which provide a type hint, that hint may be `DT_UNKNOWN`.  In cases where we do not know the type of the entry, the code tries first to unlink it as if it were something other than a directory:

https://github.com/rust-lang/rust/blob/4b043faba34ccc053a4d0110634c323f6c03765e/library/std/src/sys/unix/fs.rs#L1700-L1712

The code assumes that `unlinkat()` on a directory will fail with `EPERM`, or on Linux with the non-standard `EISDIR`, but this is not always the case.  POSIX does suggests `EPERM` as a failure in some, but not all cases:

> [EPERM]
>    The file named by path is a directory, and either the calling process does not have appropriate privileges, or the implementation prohibits using unlink() on directories.

(See [unlink](https://pubs.opengroup.org/onlinepubs/9699919799/functions/unlink.html) in _The Open Group Base Specifications Issue 7, 2018 edition_)

Implementations are not _required_ to prohibit `unlink()` on directories, and indeed UFS file systems on present day illumos and presumably at least some Solaris systems allow `unlink()` of a directory if the user has sufficient privileges.  Other platforms may as well.

Unfortunately unlinking a directory on UFS causes a sort of file system corruption that requires an unmount and a trip through `fsck` to correct.  Any code that uses `std::fs::remove_dir_all()` on UFS has since started causing that kind of corruption, which took a bit of work to track down.

I think the fix is probably relatively simple, and has the benefit of not touching the code path where we believe we know what sort of entry we are removing:

```rust
None => match cvt(unsafe { unlinkat(fd, child.name_cstr().as_ptr(), libc::AT_REMOVEDIR) }) {
    // type unknown - try to rmdir
    Err(err)
        if err.raw_os_error() == Some(libc::EEXIST)
            || err.raw_os_error() == Some(libc::ENOTEMPTY) =>
    {
        // this is a directory that contains files to remove
        remove_dir_all_recursive(Some(fd), Path::new(&child.file_name()))?;
    }
    Err(err) if err.raw_os_error() == Some(libc::ENOTDIR) {
        // this is some other type of entry that we can just unlink
        cvt(unsafe { unlinkat(fd, child.name_cstr().as_ptr(), 0) })?;
    }
    result => {
        result?;
    }
}
```

In short, we should try to `rmdir()` first, rather than `unlink()` first, as that always fails in a safe way that we can detect.

**Comment by Urgau** [comment]
cc @hkratz (as you're the one how wrote the new code)

**Comment by hkratz** [comment]
@Urgau Thanks for the ping

@rustbot claim

**Comment by apiraino** [comment]
Assigning priority as discussed in the [Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.2394335.20Fix.20for.20CVE-2022-21658.20is.20unsafe.20on.20some.20file.20systems) of the Prioritization Working Group.

@rustbot label -I-prioritize +P-critical

**Comment by nagisa** [comment]
Shouldn't this issue be ideally fixed in the filesystem itself? Filesystems corrupting themselves seem like a poor failure mode in any case, and in some ways is a violation of the sandboxing that the kernel ought to provide for the user-space applications.

**Comment by jclulow** [comment]
> Was this issue reported upstream? Filesystems corrupting themselves seem like a poor failure mode in any case, and in some ways is a violation of the sandboxing that the kernel ought to provide for the user-space applications.

It was reported upstream, and we're working on a fix -- but that fix will obviously only be in new versions of the operating system.

To expand further on the corruption: the file system becomes inconsistent and requires repair, and the damage such as it is can only be caused by privileged users; i.e., by root or a user with root equivalent privileges.  I am not yet aware of a way that it would cause actual data loss, just an interruption to service, but anything is possible.

Allowing the linking and unlinking of directories is a regrettable, but seemingly intentional feature we are saddled with through our UNIX heritage.  I'm working to disable and eventually remove the feature in the OS -- although we strongly value backwards compatibility, in this case I can't really see a legitimate use case or find a consumer that would be broken by doing so, but we won't know for sure until we try.

While this specific case is unfortunate, and we can and will work on our end to try to mitigate the impact, the fact remains that POSIX explicitly documents and allows for implementations where `unlink()` on a directory does not return `EPERM`.  For the code to be completely correct on all POSIX systems, it cannot make that assumption.

In addition, working around the issue in new versions of the OS doesn't fix existing deployments out there already.  Enterprise customers are often reluctant to promptly update their operating system, even though they may update software built and deployed on it -- software that may be written in Rust.  I myself hit this issue by upgrading only the Rust compiler to a new stable version, and trying to build and use Rust-based software we have to build UFS ramdisks on illumos.

We aim to fix the OS for the future, but the critical regression in Rust needs to be fixed as well for the present.  We appreciate everybody who has looked at this so far -- please let me know how I can help expedite the process!

**Comment by hkratz** [comment]
@jclulow Thanks again for the report.

The linked PR should fix the issue. Apparently only [Linux](https://man7.org/linux/man-pages/man2/unlink.2.html) and [FreeBSD](https://www.freebsd.org/cgi/man.cgi?query=unlink&sektion=2) _guarantee_ that trying to `unlink()` a directory fails. [Macos](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/unlink.2.html),  [NetBSD](https://man.netbsd.org/unlink.2) and [OpenBSD](https://man.openbsd.org/unlink.2) allow it, at least in theory. After looking around a bit I found that it is even explicitly documented for [Illumos](https://illumos.org/man/2/unlinkat) and [Solaris](https://docs.oracle.com/cd/E88353_01/html/E37841/unlink-2.html). So it is not a bug but a feature...

After changing that in https://github.com/illumos/illumos-gate/commit/ad8f9d956254e0caad9e4f8c85217f97cbdcade2 you might want to adapt the man page as well.

**Comment by cuviper** [comment]
> Unfortunately unlinking a directory on UFS causes a sort of file system corruption that requires an unmount and a trip through `fsck` to correct. Any code that uses `std::fs::remove_dir_all()` on UFS has since started causing that kind of corruption, which took a bit of work to track down.

We can try to avoid this case by default, as #94446 is switching to try unknown as a directory first, but you should note that this is still going to be racy. If there's an unprivileged adversary like the CVE supposes, then they can try to race changes between "try (and fail) as a directory" to "unlink as a file", to make it unlink a directory after all and induce corruption.

IOW, it sounds like it's not safe for a privileged process to unlink paths *at all* if they may be changed by lesser privileges .

**Comment by hkratz** [comment]
It should not really cause _corruption_ according to the [Illumos man page](https://illumos.org/man/2/unlinkat), just orphaned directories:
> If the path argument is a directory and the filesystem supports unlink()
>        and unlinkat() on directories, the directory is unlinked from its parent
>        with no cleanup being performed.  In UFS, the disconnected directory will
>        be found the next time the filesystem is checked with fsck(1M). The
>        unlink() and unlinkat() functions will not fail simply because a
>        directory is not empty. The user with appropriate privileges can orphan a
>        non-empty directory without generating an error message.

Of course this can lead to resource exhaustion since the disk space and the inodes cannot be reclaimed.

**Comment by cuviper** [comment]
> The user with appropriate privileges can orphan a
> non-empty directory without generating an error message.

And that user is only root or otherwise root-like?
i.e. it's not just talking about regular access privileges?

**Comment by hkratz** [comment]
> > The user with appropriate privileges can orphan a
> > non-empty directory without generating an error message.
> 
> And that user is only root or otherwise root-like? i.e. it's not just talking about regular access privileges?

For the process must have root privileges or Illumos `PRIV_SYS_LINKDIR` needs to be in the process privilege set ([Illumos privileges(5) man page](https://illumos.org/man/5/privileges)). Solaris has at some point abolished this privilege and apparently does not allow unlinking directories anymore ([Solaris privileges(5) man page](https://docs.oracle.com/cd/E36784_01/html/E36883/privileges-5.html))
